flow-server/src/main/java/com/vaadin/flow/router/RouteNotFoundError.java

```diff
@@ -68,13 +69,27 @@ public int setErrorParameter(BeforeEnterEvent event,
 
         boolean productionMode = event.getUI().getSession().getConfiguration()
                 .isProductionMode();
+        String template;
+        String routes = getRoutes(event);
+
+        if (productionMode) {
+            template = LazyInit.PRODUCTION_MODE_TEMPLATE;
+        } else if (routes.isEmpty()) {
+            // The idea of showing a different error page when there are no
+            // routes comes from application generated from start.spring.io, see
+            // https://github.com/vaadin/flow/issues/16432
+            template = readHtmlFile(EndpointRequestUtil.isHillaAvailable()
+                    ? "NoRoutesError_hilla.html"
+                    : "NoRoutesError_dev.html");
+        } else {
+            template = readHtmlFile("RouteNotFoundError_dev.html");
+        }
 
-        String template = getErrorHtml(productionMode);
         // {{routes}} should be replaced first so that it's not possible to
         // insert {{routes}} snippet via other template values which may result
         // in the listing of all available routes when this shouldn't not happen
```

Comment:
small typo not related to the PR:
```suggestion
        // in the listing of all available routes when this shouldn't happen
```